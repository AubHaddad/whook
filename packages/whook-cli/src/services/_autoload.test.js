describe('$autoload', () => {
  const WHOOK_CLI_SRC = '/var/lib/node/node_modules/@whook/cli/src';
  const log = jest.fn();
  const requireMock = jest.fn();
  const resolveMock = jest.fn();
  const $baseAutoload = jest.fn();

  beforeEach(() => {
    jest.resetModules();
    log.mockReset();
    requireMock.mockReset();
    resolveMock.mockReset();
  });

  it('should warn with bad commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      return async () => $baseAutoload;
    });
    resolveMock.mockReturnValueOnce(
      '/var/lib/node/node_modules/@whook/whook/src/index.js',
    );
    requireMock.mockReturnValueOnce({});

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      WHOOK_CLI_SRC,
      require: requireMock,
      resolve: resolveMock,
      log,
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS: ['@whook/whook'],
    });
    const { path, initializer } = await $autoload('handlerCommand');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      resolveCalls: resolveMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with project commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      return async () => $baseAutoload;
    });
    resolveMock.mockReturnValueOnce(
      '/var/lib/node/node_modules/@whook/whook/src/index.js',
    );
    requireMock.mockReturnValueOnce({
      default: async () => async () => log('warning', 'Command called!'),
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      WHOOK_CLI_SRC,
      require: requireMock,
      resolve: resolveMock,
      log,
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS: ['@whook/whook'],
    });
    const { path, initializer } = await $autoload('handlerCommand');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      resolveCalls: resolveMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with whook-cli commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      return async () => $baseAutoload;
    });
    resolveMock.mockReturnValueOnce(
      '/var/lib/node/node_modules/@whook/whook/src/index.js',
    );
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    requireMock.mockImplementationOnce(() => ({
      default: async () => async () => log('warning', 'Command called!'),
    }));

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      WHOOK_CLI_SRC,
      require: requireMock,
      resolve: resolveMock,
      log,
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS: ['@whook/whook'],
    });
    const { path, initializer } = await $autoload('handlerCommand');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      resolveCalls: resolveMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with bad commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      return async () => $baseAutoload;
    });
    resolveMock.mockReturnValueOnce(
      '/var/lib/node/node_modules/@whook/whook/src/index.js',
    );
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    $baseAutoload.mockResolvedValueOnce({
      initializer: async () => async () => {},
      path: 'mocked://service',
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      WHOOK_CLI_SRC,
      require: requireMock,
      resolve: resolveMock,
      log,
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS: ['@whook/whook'],
    });
    const { path, initializer } = await $autoload('handlerCommand');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      resolveCalls: resolveMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should delegate to whook $autoloader', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      return async () => $baseAutoload;
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      WHOOK_CLI_SRC,
      require: requireMock,
      resolve: resolveMock,
      log,
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS: ['@whook/whook'],
    });
    const { path, initializer } = await $autoload('anotherService');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      resolveCalls: resolveMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
    }).toMatchSnapshot();
  });
});
