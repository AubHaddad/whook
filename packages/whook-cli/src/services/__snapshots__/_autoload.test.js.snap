// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$autoload should delegate to whook $autoloader 1`] = `
Object {
  "baseAutoloaderCalls": Array [
    Array [
      "anotherService",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
  ],
  "path": "mocked://service",
  "requireCalls": Array [],
  "resolveCalls": Array [],
  "result": undefined,
}
`;

exports[`$autoload should warn with bad commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Plugin \\"@whook/whook\\" resolved to: /var/lib/node/node_modules/@whook/whook/src",
    ],
    Array [
      "warning",
      "The handler seems to have no default export.",
    ],
  ],
  "path": "command://handlerCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "@whook/whook",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with bad commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Plugin \\"@whook/whook\\" resolved to: /var/lib/node/node_modules/@whook/whook/src",
    ],
    Array [
      "debug",
      "Command \\"handler\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
    Array [
      "debug",
      "Command \\"handler\\" not found in: /var/lib/node/node_modules/@whook/whook/src/commands/handler",
    ],
    Array [
      "warning",
      "Command \\"handler\\" not found.",
    ],
  ],
  "path": "command://handlerCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/handler",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/cli/src/commands/handler",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "@whook/whook",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with project commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Plugin \\"@whook/whook\\" resolved to: /var/lib/node/node_modules/@whook/whook/src",
    ],
    Array [
      "warning",
      "Command called!",
    ],
  ],
  "path": "command://handlerCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "@whook/whook",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with whook-cli commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Plugin \\"@whook/whook\\" resolved to: /var/lib/node/node_modules/@whook/whook/src",
    ],
    Array [
      "debug",
      "Command \\"handler\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
    Array [
      "warning",
      "Command called!",
    ],
  ],
  "path": "command://handlerCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/handler",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/handler",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "@whook/whook",
    ],
  ],
  "result": undefined,
}
`;
